BREAKING: revert back to forceful expiration by default

Bryce-Soghigian opened the pull request:
<!-- Please follow the guidelines at https://www.conventionalcommits.org/en/v1.0.0/ and use one of the following in your title:
feat:            <-- New features that require a MINOR version update
fix:             <-- Bug fixes that require at PATCH version update
chore:           <-- Smaller changes that impact behavior but aren't large enough to be features
perf:            <-- Code changes that improve performance but do not impact behavior
docs:            <-- Documentation changes that do not impact code
test:            <-- Test changes that do not impact behavior
ci:              <-- Changes that affect test or rollout automation
!${type}:        <-- Include ! if your change includes a backwards incompatible change.
-->

Fixes #N/A <!-- issue number -->

**Description**
This PR is an implementation of the preferred implementation of the [Forceful Expiration Design](https://github.com/kubernetes-sigs/karpenter/pull/1303). 

We remove status condition 'expired', node disruption budgets for 'expired' and change the behavior to immediately delete the nodeclaim when we see a nodeclaim is past its expiry ttl.
**How was this change tested?**
- make presubmit

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

njtran commented on pkg/controllers/disruption/suite_test.go:
```diff
@@ -281,7 +281,7 @@ var _ = Describe("Simulate Scheduling", func() {
 
 		// Mark all nodeclaims as expired
```
probably want to adjust this comment too

Bryce-Soghigian replied:
```suggestion
		// Mark all nodeclaims as drifted
```

njtran commented on pkg/controllers/nodeclaim/disruption/expiration.go:
We should probably remove the status condition bits, and just simply delete the node if it's expired, and exit if it's not.

Bryce-Soghigian commented on pkg/controllers/nodeclaim/disruption/expiration.go:
```diff
@@ -36,36 +36,28 @@ type Expiration struct {
 	}
 	expirationTime := nodeClaim.CreationTimestamp.Add(*nodePool.Spec.Disruption.ExpireAfter.Duration)
-	// 2. If the NodeClaim isn't expired, remove the status condition.
+	// 2. If the NodeClaim isn't expired, remove the status condition and leave the reconcile loop
```
```suggestion
	// 2. If the NodeClaim isn't expire leave the reconcile loop.
```

Bryce-Soghigian commented on pkg/controllers/nodeclaim/disruption/expiration.go:
```diff
@@ -36,36 +36,28 @@ type Expiration struct {
-			log.FromContext(ctx).V(1).Info("removing expired status condition, not expired")
-		}
-		// If the NodeClaim isn't expired and doesn't have the status condition, return.
+		// If the NodeClaim isn't expired, return.
```
```suggestion
```

Bryce-Soghigian commented:
I removed the status conditions all together for expiration, and refactored a couple of the tests so we don't lose coverage(some of them were just using expiry as a replace action)

njtran commented on pkg/controllers/nodeclaim/disruption/expiration_test.go:
```diff
@@ -64,38 +63,30 @@ var _ = Describe("Expiration", func() {
-		Expect(nodeClaim.StatusConditions().Get(v1beta1.ConditionTypeExpired)).To(BeNil())
 	})
-	It("should mark NodeClaims as expired if the nodeClaim is expired but the node isn't", func() {
+	It("should delete NodeClaims as expired if the nodeClaim is expired but the node isn't", func() {
```
nit:
```suggestion
	It("should delete NodeClaims if the nodeClaim is expired but the node isn't", func() {
```

Bryce-Soghigian commented on pkg/controllers/nodeclaim/disruption/expiration.go:
```diff
@@ -36,36 +36,26 @@ type Expiration struct {
+		return reconcile.Result{}, err
 	}
+	// At this point we know we would have returned
+	log.FromContext(ctx).V(1).Info("deleting expired nodeclaim")
```
Agreed it would make more sense in general for us to have the termination metrics only after successful termination.

Bryce-Soghigian replied:
Added the behavior and a test validating that same behavior

njtran commented on pkg/controllers/nodeclaim/disruption/expiration.go:
```diff
@@ -36,36 +36,31 @@ type Expiration struct {
+	metrics.NodeClaimsTerminatedCounter.With(prometheus.Labels{
+		metrics.ReasonLabel:       metrics.ExpirationReason,
+		metrics.NodePoolLabel:     nodeClaim.Labels[v1beta1.NodePoolLabelKey],
+		metrics.CapacityTypeLabel: nodeClaim.Labels[v1beta1.CapacityTypeLabelKey],
```
This is super small, but we don't check the existence of this label in our controller code. I don't think the lack of capacity type label should block us from deleting this, but we should have some good null value. I was thinking something like `n/a` or `-`, so that we can aggregate on the empty value.

njtran replied:
I can see your point, but I think it's probably easiest for this PR to add the metric in here, and refactoring this to the termination controller later. I think we should add in the null value label though.

njtran reviewed (commented):
/hold
/lgtm
/approve
approving but holding so @Bryce-Soghigian can unhold when he's ready

Bryce-Soghigian commented:
/unhold
